src/Convert/ExprUtils.hs

```diff
@@ -27,16 +27,16 @@ simplify = simplifyStep . traverseSinglyNestedExprs simplify . simplifyStep
 
 simplifyStep :: Expr -> Expr
 
-simplifyStep (UniOp LogNot (Number n)) =
+simplifyStep (UniOpA LogNot a (Number n)) =
     case numberToInteger n of
         Just 0 -> bool True
         Just _ -> bool False
-        Nothing -> UniOp LogNot $ Number n
-simplifyStep (UniOp LogNot (BinOp Eq a b)) = BinOp Ne a b
-simplifyStep (UniOp LogNot (BinOp Ne a b)) = BinOp Eq a b
+        Nothing -> UniOpA LogNot a $ Number n
+simplifyStep (UniOpA LogNot a (BinOpA Eq a' l r)) = BinOpA Ne (a <> a') l r
+simplifyStep (UniOpA LogNot a (BinOpA Ne a' l r)) = BinOpA Eq (a <> a') l r
```

Comment:
I'm hesitant to perform simplifications on expressions with attributes without having clear understanding of the user intent. My understanding is that these attributes are typically used in cases where the designer is hoping to communicate something very specific to the synthesis tool. Will users really want us to transform their carefully-chosen expressions in these ways?

I think none of the simplifications performed in this file are necessary for _correctness_, so it seems like we should be conservative and not try too hard to optimize expressions that contain attributes.

I actually think most of your changes in this file look good, but I am specifically unsure of instances where we are transferring an attribute onto a different operator (or in a different context) or concatenating two sets of attributes.

Reply:
The reasoning I had was 1: preserving attributes and 2: moving attributes to follow the non constant parts of the expressions. I followed the same rules for binary expressions.